### PR TITLE
[4.0] RavenDB-10581 when we are using / in the ID, then we need to check if the generated ID does not overlap already existing one

### PR DESCRIPTION
**Pull request by ppekrol**

### CODE REVIEW

File: src/Raven.Server/Documents/TransactionOperationsMerger.cs [hunk 1]
```diff
@@ -60,7 +62,12 @@ public TransactionOperationsMerger(DocumentDatabase parent, CancellationToken sh
 
         public void Start()
         {
-            _txLongRunningOperation = PoolOfThreads.GlobalRavenThreadPool.LongRunning(x => MergeOperationThreadProc(), null, TransactionMergerThreadName);            
+            _txLongRunningOperation = PoolOfThreads.GlobalRavenThreadPool.LongRunning(x => MergeOperationThreadProc(), null, TransactionMergerThreadName);
+        }
+
+        public abstract class DocumentPutTransactionCommand : MergedTransactionCommand
+        {
+            public bool CheckIfGeneratedIdIsNotOverlapping;
```
Comment (ayende) [inline src/Raven.Server/Documents/TransactionOperationsMerger.cs]: Why, this is relevant only on PUT on new docs. Not relevant anywhere else (`PATCH`, for example)

File: src/Raven.Server/Documents/DocumentPutAction.cs [hunk 2]
```diff
@@ -301,11 +316,31 @@ private string BuildDocumentId(string id, long newEtag, out bool knownNewId)
                         i += 19;
                         valuePtr[i] = '-';
 
-                        Format.Backwards.WriteNumber(valuePtr + i - 1, (ulong)newEtag);
+                        if (checkIfGeneratedIdIsNotOverlapping == false)
+                            Format.Backwards.WriteNumber(valuePtr + i - 1, (ulong)newEtag);
+                        else
+                        {
+                            while (true)
```
Comment (ayende) [inline src/Raven.Server/Documents/DocumentPutAction.cs]: I don't want this solution, this does a scan, and can get into bad situation.
Just change the generated doc id in this case.

File: src/Raven.Server/Documents/DocumentPutAction.cs [hunk 3]
```diff
@@ -170,7 +172,17 @@ public DocumentPutAction(DocumentsStorage documentsStorage, DocumentDatabase doc
...
                     {
-                        table.Insert(tvb);
+                        try
+                        {
+                            table.Insert(tvb);
+                        }
+                        catch (VoronConcurrencyErrorException)
+                        {
+                            if (knownNewId && checkIfGeneratedIdIsNotOverlapping == false)
+                                ThrowOverlappingGeneratedNewIdException(id);
```
Comment (ayende) [inline src/Raven.Server/Documents/DocumentPutAction.cs]: This here should recurse immediately and try (once).

**Review by ayende — CHANGES_REQUESTED** [review]

**Comment by ppekrol** [comment]
LGTM

**Comment by arekpalinski** [comment]
LGTM, although " Server compilation failed. "

**Comment by ppekrol** [comment]
I will fix that

**Pull request merged**